code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/SmallImageTemplateEventHandler.kt

```diff
@@ -32,7 +34,25 @@ internal class SmallImageTemplateEventHandler(private val callback: ContentCardC
         if (event is UIEvent.Interact) {
             onInteractEvent(event, propositionId)
         } else {
-            super.onEvent(event, propositionId)
+            if (event.aepUi is SmallImageUI) {
+                val smallImageUI = event.aepUi as SmallImageUI
+                val currentUiState = event.aepUi.getState() as SmallImageCardUIState
+
+                // For dismiss event, change the dismissed state of the UI to true
+                // so that the UI can be removed from the screen
+                if (event is UIEvent.Dismiss) {
+                    smallImageUI.updateState(currentUiState.copy(dismissed = true))
+                }
+
+                // Call the super class to call the callback and send tracking event if needed
+                super.onEvent(event, propositionId)
+
+                // For display event, change the displayed state of the UI to true
+                // after the tracking event is sent for the initial composition
+                if (event is UIEvent.Display) {
+                    smallImageUI.updateState(currentUiState.copy(displayed = true))
+                }
+            }
```

Comment:
- Is there a reason this particular ordering is needed for sending events? Can we not call super for all events?
- Given that this class deals with SmallImageUi only, we can cast it eagerly and bail if it is unrelated to SmallImageUi allowing you to simplify this function

```
  val smallImageUI : SmallImageUI = event.aepUi as? SmallImageUI ?: return
  val currentUiState = smallImageUI.getState() as SmallImageCardUIState
   when (event) {
          is UIEvent.Dismiss -> {
              smallImageUI.updateState(currentUiState.copy(dismissed = true))
          }
          is UIEvent.Display -> {
              smallImageUI.updateState(currentUiState.copy(displayed = true))
          }

          is UIEvent.Interact -> {
              onInteractEvent(event, propositionId)
          }

        super.onEvent(event, propositionId)
    }
}
```

Reply:
If `displayed = true` is set to true before calling `super.onEvent`, the tracking event, which is only sent when `displayed` is false, will never work. 
Will make the change for casting early 👍

Reply:
So calling super before changing the state should work right?

Reply:
Changed to have the super class `onEvent` not be open. Instead, provided an abstract function `handleEvent` to be implemented by individual template listeners. Base class `MessagingEventHandler.onEvent` will always be called first now. It can perform the tasks common to all templates and delegate the specific tasks to sub classes by calling `handleEvent`.

cc: @rymorale